downstream.go

```diff
@@ -1510,6 +1524,116 @@ func (dc *downstreamConn) handleMessageRegistered(msg *irc.Message) error {
 			Command: "INVITE",
 			Params:  []string{upstreamUser, upstreamChannel},
 		})
+	case "CHATHISTORY":
+		var subcommand string
+		if err := parseMessageParams(msg, &subcommand); err != nil {
+			return err
+		}
+		var target, criteria, limitStr string
+		if err := parseMessageParams(msg, nil, &target, &criteria, &limitStr); err != nil {
+			return ircError{&irc.Message{
+				Command: "FAIL",
+				Params:  []string{"CHATHISTORY", "NEED_MORE_PARAMS", subcommand, "Missing parameters"},
+			}}
+		}
+
+		if dc.srv.LogPath == "" {
+			return ircError{&irc.Message{
+				Command: irc.ERR_UNKNOWNCOMMAND,
+				Params:  []string{dc.nick, subcommand, "Unknown command"},
+			}}
+		}
+
+		uc, entity, err := dc.unmarshalEntity(target)
+		if err != nil {
+			return err
+		}
+
+		// TODO: support msgid criteria
+		criteriaParts := strings.SplitN(criteria, "=", 2)
+		if len(criteriaParts) != 2 || criteriaParts[0] != "timestamp" {
+			return ircError{&irc.Message{
+				Command: "FAIL",
+				Params:  []string{"CHATHISTORY", "UNKNOWN_CRITERIA", criteria, "Unknown criteria"},
+			}}
+		}
+
+		timestamp, err := time.Parse(serverTimeLayout, criteriaParts[1])
+		if err != nil {
+			return ircError{&irc.Message{
+				Command: "FAIL",
+				Params:  []string{"CHATHISTORY", "INVALID_CRITERIA", criteria, "Invalid criteria"},
+			}}
+		}
+
+		limit, err := strconv.Atoi(limitStr)
+		if err != nil || limit < 0 || limit > dc.srv.HistoryLimit {
+			return ircError{&irc.Message{
+				Command: "FAIL",
+				Params:  []string{"CHATHISTORY", "INVALID_LIMIT", limitStr, "Invalid limit"},
+			}}
+		}
+
+		switch subcommand {
+		case "BEFORE":
+			batchRef := "history"
+			dc.SendMessage(&irc.Message{
+				Prefix:  dc.srv.prefix(),
+				Command: "BATCH",
+				Params:  []string{"+" + batchRef, "chathistory", target},
+			})
+
+			history := make([]historyMessage, limit)
+			remaining := limit
+
+			first := true
+			tries := 0
+			for remaining > 0 {
+				buf, err := parseMessages(uc.network, entity, timestamp, remaining, first)
```

Comment:
Why do we need the first call to behave differently?

Reply:
For the first call, we only want messages up to a certain hour, minute, second. For the next calls (the previous days), we want messages up to the end of that day. (In other words, if we're searching before Jan 2 15:04:05, then the call for Jan 2 will search until 15:04:05, whereas thel calls for Jan 1, Dec 31, ... will search until midnight.)

Reply:
What about replacing this:

```go
timestamp = timestamp.AddDate(0, 0, -1)
```

with this:

```go
t = time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, t.Location()).Add(-1)
```

That way, no need for a special case.

Reply:
This is simpler, but it would incorrectly skip over messages posted on the very last nanosecond of that day. Which solution should I pick?

Reply:
I think this trade-off is reasonable. Simplicity, correctness, performance, in this order.

Reply:
Makes sense. I'll implement the simple solution then.